lib/plugins/cpuUsageThrottle.js

```diff
@@ -0,0 +1,139 @@
+    assert.object(opts, 'opts');
+    assert.number(opts.limit, 'opts.limit');
+    assert.optionalNumber(opts.max, 'opts.max');
+    assert.number(opts.interval, 'opts.interval');
+    assert.optionalNumber(opts.halfLife, 'opts.halfLife');
+
+    if (opts.err !== undefined && opts.err !== null) {
+        assert.ok(opts.err instanceof Error, 'opts.err must be an error');
+        assert.optionalNumber(opts.err.statusCode, 'opts.err.statusCode');
+    }
+
+    var self = {};
+    self._err = opts.err || defaultResponse;
+    self._limit = opts.limit;
+    self._max = opts.max || 1;
+    self._interval = opts.interval;
+    self._halfLife = (typeof opts.halfLife === 'number') ?
+      opts.halfLife : opts.interval;
+    self._timeout = null;
+    assert.ok(self._max > self._limit, 'limit must be less than max');
+
+    // We start at 0 to compensate for the initial spike in CPU usage a server
+    // experiences when starting up.
+    self._ewma = new EWMA(self._halfLife, 0);
+
+    // self._reject represents the % of traffic that we should reject at the
+    // current point in time based on how much over our limit we are. This is
+    // updated on an interval by updateReject().
+    self._reject = 0;
+
+    // updateReject should be called on an interval, it checks the average CPU
+    // usage between two invocations of updateReject.
+    function updateReject() {
+        pidusage.stat(process.pid, function (e, stat) {
+            // If we fail to stat for whatever reason, err on the side of
+            // accepting traffic, while failure probably indicates something
+            // horribly wrong with the server, identifying this error case and
+            // appropriately responding is way outside the scope of this plugin
+            if (e) {
+                // When cpu is NaN, we should never reject
+                stat = { cpu: NaN };
+            }
+
+            // Divide by 100 to match linux's top format
+            self._cpu = stat.cpu / 100;
+            // Update reject with the % of traffic we should be rejecting. This
+            // is safe since max > limit so it can never be division by 0. If
+            // the current cpu usage is less that the limit, _reject will be
+            // negative and we will never shed load
+            self._reject =
+                (self._cpu - self._limit) / (self._max - self._limit);
+            self._timeout = setTimeout(updateReject, self._interval);
+        });
+    }
+
+    // Kick off updating our _reject value
+    updateReject();
+
+    function onRequest (req, res, next) {
+        // Check to see if this request gets rejected
+        if (self._reject > Math.random()) {
+            // If so, log some metadata and return an error
+            req.log.trace({
+                plugin: 'cpuUsageThrottle',
+                cpuUsage: self._cpu,
+                limit: self._limit
+            }, 'maximum cpu usage exceeded, rejecting request');
+            res.send(self._err);
```

Comment:
When working with errors, let's use `return next(self._err)`

Reply:
This is so the err events emitted by server will work correctly, but should also make your tests a little easier so you don't have to mock res.send

Reply:
Oh cool, didn't know `next` handled sending when an error was passed in! https://github.com/restify/node-restify/blob/master/lib/server.js#L925-L926